### PR TITLE
Improve error handling

### PR DESCRIPTION
**Pull request by christianalfoni**
I have changed the way we manage errors. Now the `api` throws errors as normal. In our actions logic we catch these errors (already do so) and pass them into an action called `actions.internal.handleError(error)`. This action evaluates what kind of error was thrown and gives the proper indication to the user.

That means any `try/catch` in our code should pass its error to `actions.internal.handleError`.

The reason it works this way is because we want to catch errors in our actions logic, but we do not want to evaluate per catch if the error is already handled by the API error logic or a custom user indication needs to be produced.

### CODE REVIEW

**Comment by lbogdan** [comment]
Build for latest commit https://github.com/codesandbox/codesandbox-client/pull/3089/commits/939039c91f5228db06041679b7020f1fca8213f3 is at https://pr3089.build.csb.dev/s/new.

**Comment by CompuIves** [comment]
I'm thinking, would it be valuable to also show the user the actual error message that comes from the API? That might help us with error reports coming in.

**Comment by christianalfoni** [comment]
Okay, updated now so that we pass `{ message: 'custom message', error }`. It looks like this:

Screenshot 2019-12-11 at 10 14 58

File: packages/app/src/app/overmind/effects/executor.ts [hunk 1]
```diff
@@ -1,9 +1,19 @@
 import { Sandbox } from '@codesandbox/common/lib/types';
 import { executorsManager } from 'app/utils/executor-manager';
 
+let _isInitialized = false;
```
Comment (CompuIves) [inline packages/app/src/app/overmind/effects/executor.ts]: Some unintended changes creeped in?

File: packages/app/src/app/overmind/factories.ts [hunk 2]
```diff
@@ -41,31 +38,10 @@ export const withLoadApp = <T>(
...
-            ],
-          },
-        });
-      } else {
-        effects.notificationToast.error(
-          "We weren't able to sign you in, this could be due to a flaky connection or something on our server. Please try again in a minute."
-        );
-      }
+      actions.internal.handleError({
+        message: 'Could not authorize user',
```
Comment (CompuIves) [inline packages/app/src/app/overmind/factories.ts]:
```suggestion
        message: 'We had trouble with signing you in',
```

Comment (CompuIves) [reply packages/app/src/app/overmind/factories.ts]: Less 🤖 

😄

File: packages/app/src/app/overmind/internalActions.ts [hunk 3]
```diff
@@ -349,21 +358,52 @@ export const closeTabByIndex: Action<number> = ({ state }, tabIndex) => {
   state.editor.tabs.splice(tabIndex, 1);
 };
 
-export const onApiError: Action<ApiError> = (
-  { state, actions, effects },
-  error
-) => {
-  const { response } = error;
+export const handleError: Action<{
+  message: string;
```
Comment (CompuIves) [inline packages/app/src/app/overmind/internalActions.ts]:
```suggestion
  /**
  * The message that will show as title of the notification
  ** /
  message: string;
```

File: packages/app/src/app/overmind/namespaces/editor/actions.ts [hunk 4]
```diff
@@ -260,9 +260,10 @@ export const saveClicked: AsyncAction = withOwnedSandbox(
           state.editor.changedModuleShortids.push(moduleShortid);
         }
       });
-      effects.notificationToast.error(
-        'Sorry, was not able to save, please try again'
-      );
+      actions.internal.handleError({
+        message: 'Sorry, was not able to save, please try again',
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/editor/actions.ts]:
```suggestion
        message: 'There was a problem with saving the files, please try again',
```

File: packages/app/src/app/overmind/namespaces/editor/internalActions.ts [hunk 5]
```diff
@@ -168,7 +168,10 @@ export const saveCode: AsyncAction<{
       tab.dirty = false;
     }
   } catch (error) {
-    effects.notificationToast.warning(error.message);
+    actions.internal.handleError({
+      message: 'Unable to save the code',
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/editor/internalActions.ts]:
```suggestion
      message: 'There was a problem with saving the code, please try again',
```

File: packages/app/src/app/overmind/namespaces/explore/actions.ts [hunk 6]
```diff
@@ -53,20 +55,22 @@ export const pickSandboxModal: Action<PickedSandboxDetails> = (
...
   id
 ) => {
   try {
     state.explore.selectedSandbox = await effects.api.getSandbox(id);
   } catch (error) {
-    effects.notificationToast.error(
-      'There has been a problem getting the sandbox'
-    );
+    actions.internal.handleError({
+      message: 'There has been a problem getting the sandbox',
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/explore/actions.ts]:
```suggestion
      message: 'There has been a problem fetching the sandbox',
```

File: packages/app/src/app/overmind/namespaces/explore/actions.ts [hunk 7]
```diff
@@ -79,9 +83,10 @@ export const pickedSandboxesMounted: AsyncAction = async ({
     );
     state.explore.pickedSandboxes = pickedSandboxes;
   } catch (error) {
-    effects.notificationToast.error(
-      'There has been a problem getting the sandboxes'
-    );
+    actions.internal.handleError({
+      message: 'There has been a problem getting the sandboxes',
```
Comment (CompuIves) [inline packages/app/src/app/overmind/namespaces/explore/actions.ts]:
```suggestion
      message: 'A problem occurred while fetching the sandboxes',
```

File: packages/app/src/app/overmind/effects/executor.ts (hunk 1 above)
Comment (christianalfoni) [reply packages/app/src/app/overmind/effects/executor.ts]:
```suggestion

```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 8]
```diff
@@ -1,9 +1,18 @@
 import { Sandbox } from '@codesandbox/common/lib/types';
 import { executorsManager } from 'app/utils/executor-manager';
 
+
 export default {
-  initializeExecutor(sandbox: Sandbox) {
-    return executorsManager.initializeExecutor(sandbox);
+  async initializeExecutor(sandbox: Sandbox) {
+    if (_isInitialized) {
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion

```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 9]
```diff
@@ -1,9 +1,18 @@
...
+      await this.closeExecutor();
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion

```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 10]
```diff
@@ -1,9 +1,18 @@
...
+    }
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion

```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 11]
```diff
@@ -1,9 +1,18 @@
...
+
+    const result = await executorsManager.initializeExecutor(sandbox);
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion
  return executorsManager.initializeExecutor(sandbox);
```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 12]
```diff
@@ -1,9 +1,18 @@
...
+
+    _isInitialized = true;
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion

```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 13]
```diff
@@ -1,9 +1,18 @@
...
+
+    return result;
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion

```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 14]
```diff
@@ -1,9 +1,18 @@
 import { Sandbox } from '@codesandbox/common/lib/types';
 import { executorsManager } from 'app/utils/executor-manager';
 
+
 export default {
-  initializeExecutor(sandbox: Sandbox) {
-    return executorsManager.initializeExecutor(sandbox);
+  async initializeExecutor(sandbox: Sandbox) {
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion
  initializeExecutor(sandbox: Sandbox) {
```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 15]
```diff
@@ -1,9 +1,13 @@
 import { Sandbox } from '@codesandbox/common/lib/types';
 import { executorsManager } from 'app/utils/executor-manager';
 
+
 export default {
   initializeExecutor(sandbox: Sandbox) {
-    return executorsManager.initializeExecutor(sandbox);
+
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion
```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 16]
```diff
@@ -1,9 +1,13 @@
...
+  return executorsManager.initializeExecutor(sandbox);
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion
    return executorsManager.initializeExecutor(sandbox);
```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 17]
```diff
@@ -1,9 +1,13 @@
...
+
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion
```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 18]
```diff
@@ -1,9 +1,13 @@
...
+
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion
```

File: packages/app/src/app/overmind/effects/executor.ts [hunk 19]
```diff
@@ -1,6 +1,7 @@
 import { Sandbox } from '@codesandbox/common/lib/types';
 import { executorsManager } from 'app/utils/executor-manager';
 
+
```
Comment (christianalfoni) [inline packages/app/src/app/overmind/effects/executor.ts]:
```suggestion
```

**Pull request merged**